db/db.go

```diff
@@ -0,0 +1,260 @@
+	// TODO: DeleteEntity(string) error
+
+	// Managing labels for a single entity
+	// TODO: ReadEntityLabels(string)
+	// TODO: DeleteEntityLabel(string, string)
+
+	// Managing Multiple entities
+	CreateEntities([]Entity) []string
+	ReadEntities(query.Query) ([]Entity, error)
+	UpdateEntities(query.Query, Entity) ([]Entity, error)
+	DeleteEntities(query.Query) ([]Entity, error)
+
+	// Managing connection to DB
+	Connect() error
+	Disconnect()
+}
+
+// mongo struct stores all info needed to connect and query a mongo instance.
+type mongo struct {
+	url        string
+	database   string
+	collection string
+	session    *mgo.Session
+}
+
+// NewConnector creates an instance of Connector.
+func NewConnector(url string, database string, collection string) Connector {
+	return &mongo{
+		url:        url,
+		database:   database,
+		collection: collection,
+	}
+}
+
+// Connect creates a session to db. All queries to DB will copy this session to
+// make a request (https://godoc.org/gopkg.in/mgo.v2#session.Copy).
+func (m *mongo) Connect() error {
+	session, err := mgo.Dial(m.url)
+	if err != nil {
+		return ErrConnect
+	}
+
+	m.session = session
+	return nil
+}
+
+// Disconnect must be called when finished with db connector.
+func (m *mongo) Disconnect() {
+	m.session.Close()
+}
+
+// CreateEntities inserts many Entity into DB.
+func (m *mongo) CreateEntities(entities []Entity) []string {
+	s := m.session.Copy()
+	defer s.Close()
+
+	c := m.session.DB(m.database).C(m.collection)
+
+	var objectIDs []string
+
+	for _, e := range entities {
+		objectID := bson.NewObjectId().String()
+		objectIDs = append(objectIDs, objectID)
+		// Note: this modifies entity in memory
+		e["_id"] = objectID
+
+		c.Insert(e)
+	}
+
+	return objectIDs
+}
+
+// Query mongoDB instance with a Query object, which is tranlsated to bson.M
+// object for mgo driver to understand. Returns a list of Entity objects. If
+// nothing is found, returns empty list.
+func (m *mongo) ReadEntities(q query.Query) ([]Entity, error) {
+	mgoQuery := translateQuery(q)
+
+	s := m.session.Copy()
+	defer s.Close()
+
+	var entities []Entity
+	c := s.DB(m.database).C(m.collection)
+	err := c.Find(mgoQuery).All(&entities)
+	if err != nil {
+		return nil, ErrNotFound
+	}
+
+	return entities, nil
+}
+
+// UpdateEntities does a db.collection.findAndModify() operation
+// (https://docs.mongodb.com/manual/reference/method/db.collection.findAndModify/)
+// on all docs returned from query. Returns diff of entities (old values for
+// fields changed, in addition to including "_id" field/value) or an error, if
+// one occured.
+func (m *mongo) UpdateEntities(q query.Query, u Entity) ([]Entity, error) {
+	mgoQuery := translateQuery(q)
+
+	s := m.session.Copy()
+	defer s.Close()
+
+	c := s.DB(m.database).C(m.collection)
+
+	// List of IDs for docs to update
+	ids, err := idsForQuery(c, mgoQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	// Change to make
+	change := mgo.Change{Update: bson.M{"$set": u}}
```

Comment:
Best to be explicit so our intention is clear (to our future selves and other devs) and to guard against defaults changing. So `ReturnNew: false` and code comment that Mongo returns the original doc before modifications. Speaking of which: "orig" might be more clear than "diff" because it's not really a diff.